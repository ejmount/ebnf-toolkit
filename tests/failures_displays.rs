use ebnf_toolkit::EbnfError;
use ebnf_toolkit::Rule;
use insta::assert_compact_debug_snapshot;

#[test]
fn incomplete_rule() {
    let src = "Foo = A|";
    let err = Rule::new(src).unwrap_err();

    println!("{err}");
    insta::assert_snapshot!(err);
}

#[test]
fn invalid_syntax_rule() {
    let srcs = ["Foo = A|;", "Foo = (A;", "Foo = (?;"];
    for src in srcs {
        let err = Rule::new(src).unwrap_err();

        println!("{err}");
        insta::assert_snapshot!(err);
    }
}

#[test]
fn invalid_start() {
    let srcs = [
        "'Hello' = A;",
        "A? = A;",
        "A* = A;",
        "? = A;",
        "#'aaa' = A;",
        "A|B = A;",
    ];
    for src in srcs {
        let err = Rule::new(src).unwrap_err();

        assert_eq!(err, err); // Custom implementation of PartialEq needs tested
        println!("{err}");
        insta::assert_snapshot!(err);
    }
}

#[test]
fn empty_input() {
    let src = "";
    let err = Rule::new(src).unwrap_err();

    println!("{err}");
    assert_eq!(err, EbnfError::EmptyInput);
}

#[test]
fn unclosed_string() {
    let input = "'Hello";

    let err = Rule::new(input).unwrap_err();

    println!("{err}");
    assert_eq!(err, err);
    assert_compact_debug_snapshot!(err);
}
