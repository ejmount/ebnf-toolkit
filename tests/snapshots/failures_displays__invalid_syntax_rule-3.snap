---
source: tests/failures_displays.rs
expression: err
---
[31mError:[0m 
   [38;5;246mâ•­[0m[38;5;246mâ”€[0m[38;5;246m[[0m <input>:1:1 [38;5;246m][0m
   [38;5;246mâ”‚[0m
 [38;5;246m1 â”‚[0m [38;5;249mR[0m[38;5;249mu[0m[38;5;249ml[0m[38;5;249me[0m[38;5;249m [0m[38;5;249m=[0m[38;5;249m [0m[38;5;155m([0m[38;5;187m?[0m[38;5;201m;[0m
 [38;5;240m  â”‚[0m        [38;5;155mâ”¬[0m[38;5;187mâ”¬[0m[38;5;201mâ”¬[0m  
 [38;5;240m  â”‚[0m        [38;5;155mâ•°[0m[38;5;155mâ”€[0m[38;5;155mâ”€[0m[38;5;155mâ”€[0m[38;5;155mâ”€[0m Possible unclosed bracket
 [38;5;240m  â”‚[0m         [38;5;187mâ”‚[0m[38;5;201mâ”‚[0m  
 [38;5;240m  â”‚[0m         [38;5;187mâ•°[0m[38;5;187mâ”€[0m[38;5;187mâ”€[0m[38;5;187mâ”€[0m Could not apply to preceding term
 [38;5;240m  â”‚[0m          [38;5;201mâ”‚[0m  
 [38;5;240m  â”‚[0m          [38;5;201mâ•°[0m[38;5;201mâ”€[0m[38;5;201mâ”€[0m Rule ending here did not parse successfully
 [38;5;240m  â”‚[0m 
 [38;5;240m  â”‚[0m [38;5;115mNote[0m: The parse stack looked like this (most recent on top):
 [38;5;240m  â”‚[0m       â””â”€0: UnparsedOperator [1:9..1:10]
 [38;5;240m  â”‚[0m         â”‚  â””â”€ Terminator
 [38;5;240m  â”‚[0m         1: UnparsedOperator [1:8..1:9]
 [38;5;240m  â”‚[0m         â”‚  â””â”€ Optional
 [38;5;240m  â”‚[0m         2: UnparsedOperator [1:7..1:8]
 [38;5;240m  â”‚[0m         â”‚  â””â”€ OpenedGroup
 [38;5;240m  â”‚[0m         3: UnparsedOperator [1:5..1:6]
 [38;5;240m  â”‚[0m         â”‚  â””â”€ Equals
 [38;5;240m  â”‚[0m         4: Nonterminal [1:0..1:4]
 [38;5;240m  â”‚[0m            â””â”€ Rule
[38;5;246mâ”€â”€â”€â•¯[0m
